eval/langfuse_ragas.py

```diff
@@ -0,0 +1,179 @@
+        conversation.append(message)
+    return conversation
+
+
+def traces_to_samples(
+    traces: FetchTracesResponse,
+    filter: Optional[Dict] = None,
+    return_dataset: bool = False,
+    dataset_path: Optional[str] = None,
+) -> List[BaseSample]:
+    """Convert a list of traces returned from the Langfuse API into RAGAS samples"""
+    # traces are currently deliberately only suitable to be converted into SingleTurnSample objects
+    # need to think about MultiTurnSample objects as well
+
+    samples = []
+
+    traces = traces.data
+    if filter:  # filter by user_id, e.g. filter = {'user_id':'helen'} returns only traces with user id helen
+        traces = [trace for trace in traces if all(getattr(trace, attr) == value for attr, value in filter.items())]
+
+    traces = sorted(traces, key=lambda trace: trace.timestamp)  # sort so have oldest first – this is necessary for
+    # input_appears_in_next_trace_chat_history to work
+
+    answer_history = []
+    for i, trace in enumerate(traces):
+
+        is_single_turn_sample = False
+
+        next_trace = traces[i + 1] if i < len(traces) - 1 else None
+        if next_trace:
+            input_appears_in_next_trace_chat_history = any(
+                message["content"] == trace.input["input"] for message in next_trace.input["chat_history"]
+            )
+        else:
+            input_appears_in_next_trace_chat_history = False
+
+        if input_appears_in_next_trace_chat_history:
+            # if input_appears_in_next_trace_chat_history = True, the message will eventually appear in
+            # the conversation of a MultiTurnSample, so do not instantiate a sample for it
+            # but do retain the answer history for when the MultiTurnSample is eventually instantiated
+            answer_history.append(trace.output["answer"])
+        else:
+            # trace is either a SingleTurnSample or the last trace defining a MultiTurnSample
+            is_single_turn_sample = len(trace.input["chat_history"]) == 1
+
+            if is_single_turn_sample:
+                sample = SingleTurnSample(
+                    user_input=trace.input["input"],
+                    retrieved_contexts=[context["page_content"] for context in trace.output["context"]],
+                    response=trace.output["answer"],
+                )
+            else:
+                sample = MultiTurnSample(
+                    user_input=trace_to_conversation(
+                        trace, answer_history
+                    ),  # contexts are not passed in to MultiTurnSample objects
+                )
+            samples.append(sample)
+            answer_history = []
+
+    if return_dataset or dataset_path:
+        dataset = EvaluationDataset(samples=samples)
+        if dataset_path:
+            dataset_path.to_csv(dataset_path)
+
+        if return_dataset:
+            return dataset
+
+    return samples
+
+
+if __name__ == "__main__":
+
+    traces = langfuse.fetch_traces()
```

Comment:
Just curious, at the moment it would fetch all traces belonging to the nesta brain langfuse project?

Reply:
Yes, so we might want to filter them later. There aren't many at the moment.